docs/dev/writing-code.md

```diff
@@ -0,0 +1,209 @@
+# Writing code in aws-crt-python
+
+`aws-crt-python` provides "language bindings", allowing Python to use the
+C libraries which make up the AWS SDK Common Runtime (CRT).
+
+This is not easy code to write. You must know Python. You must know C.
+You must learn how the aws-c libraries do error handling and memory management,
+you must learn how the Python C-API does error handling and memory management,
+and you must mix the two styles together. This code is multithreaded and asynchronous.
+Buckle up.
+
+### Table of Contents
+
+*   [Required Reading](#required-reading)
+*   [Writing Python Code](#writing-python-code)
+    *   [General](#general-python-rules)
+        *   [Naming Conventions](#python-naming-conventions)
+        *   [Use Type Hints](#use-type-hints)
+    *   [Forward and Backward Compatibility](#forward-and-backward-compatibility)
+        *   [Functions with Lots of Arguments](#functions-with-lots-of-arguments)
+        *   [Use `None` for Optional Arguments](#use-none-as-the-default-value-for-optional-arguments)
+        *   [Callback Signatures](#callback-signatures)
+        *   [Be Careful When Adding](#be-careful-when)
+    *   [Asynchronous APIs](#asynchronous-apis)
+*   [Lifetime Management](#lifetime-management)
+*   [Writing C Code](#writing-c-code)
+
+## Required Reading
+
+*   `Developer Guide for the aws-c libraries` - (TODO: WRITE THIS)
```

Comment:
Can we link the section [here](https://github.com/awslabs/aws-c-common#coding-style)? Or update that section and it can be shared across aws-c-* and other aws-crt-*.